server/src/main/java/org/elasticsearch/cluster/coordination/StableMasterHealthIndicatorService.java

```diff
@@ -0,0 +1,170 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.cluster.coordination;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.cluster.node.DiscoveryNode;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.discovery.DiscoveryModule;
+import org.elasticsearch.health.HealthIndicatorDetails;
+import org.elasticsearch.health.HealthIndicatorImpact;
+import org.elasticsearch.health.HealthIndicatorResult;
+import org.elasticsearch.health.HealthIndicatorService;
+import org.elasticsearch.health.HealthStatus;
+import org.elasticsearch.health.ImpactArea;
+import org.elasticsearch.health.SimpleHealthIndicatorDetails;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ExecutionException;
+
+import static org.elasticsearch.health.ServerHealthComponents.CLUSTER_COORDINATION;
+
+public class StableMasterHealthIndicatorService implements HealthIndicatorService {
+
+    public static final String NAME = "stable_master";
+
+    private final ClusterService clusterService;
+    private final DiscoveryModule discoveryModule;
+    private final MasterHistoryService masterHistoryService;
+    private static final Logger logger = LogManager.getLogger(StableMasterHealthIndicatorService.class);
+
+    public StableMasterHealthIndicatorService(
+        ClusterService clusterService,
+        DiscoveryModule discoveryModule,
+        MasterHistoryService masterHistoryService
+    ) {
+        this.clusterService = clusterService;
+        this.discoveryModule = discoveryModule;
+        this.masterHistoryService = masterHistoryService;
+    }
+
+    @Override
+    public String name() {
+        return NAME;
+    }
+
+    @Override
+    public String component() {
+        return CLUSTER_COORDINATION;
+    }
+
+    @Override
+    public HealthIndicatorResult calculate(boolean includeDetails) {
+        HealthStatus stableMasterStatus;
+        String summary;
+        Map<String, Object> details = new HashMap<>();
+        Collection<HealthIndicatorImpact> impacts = new ArrayList<>();
+        MasterHistory localMasterHistory = masterHistoryService.getLocalMasterHistory();
+        if (hasSeenMasterInLast30Seconds()) {
+            logger.trace("Have seen a master in the last 30 seconds");
+            Set<DiscoveryNode> mastersInLast30Minutes = localMasterHistory.getDistinctMastersSeen();
+            if (mastersInLast30Minutes.size() >= 3) {
```

Comment:
I believe this should be `>3`

Reply:
Oh you're right -- I'd read that as `3 or more times` but it's actually `more than 3`. Fixing it (although this is not part of this PR and will be removed).

Reply:
The logic is wrong there anyway. We want to check for more than 3 changes of master, not more than 3 unique masters. I'll change that logic.

Reply:
Done.